### PR TITLE
require c++14

### PR DESCRIPTION
**Pull request by trws**
This one is short and sweet for now.  It requires C++14 but uses no features yet.  I'll go through shortly and change a couple of things to remove their conditional compilation and start directly using 14 in a few places.  Mainly unconditionally use constexpr in some places and get rid of some of the unnecessary type calculations.

### CODE REVIEW

File: include/RAJA/config.hpp.in [hunk 1]
```diff
@@ -31,6 +31,11 @@
 #ifndef RAJA_config_HPP
 #define RAJA_config_HPP
 
+static_assert(__cplusplus >= 201402L,
```
Comment (davidbeckingsale) [inline include/RAJA/config.hpp.in]: This is cool 👍

**Comment by rhornung67** [comment]
@trws will this *require* users to use C++14?

**Comment by trws** [comment]
@rhornung67: Yup, that was the idea.  It will not require them to change a single character of their code, but it will require that their build use 14 or higher as the standard being targeted.  We can always go more piecemeal, but that will mean we can't use any feature we can't macro-off to be backwards compatible with 11, so if this is an option it would make life easier (and the new camp PR much less macro-y).

**Comment by rhornung67** [comment]
I agree that there are really good reasons for the move, but there is one major concern. Please see my MS Teams message to you.

**Comment by rhornung67** [comment]
@trws do you have time to update this PR so we can merge it before the release in early January?

**Comment by trws** [comment]
I think the answer to that is yes, it needs to wait until after the OSDI deadline next week but shouldn't be too bad once I can get to it.

**Comment by rhornung67** [comment]
@trws do you mind if I pull your fork into the RAJA repo for a new PR referencing this one so we can run the Gitlab CI more easily?

**Comment by trws** [comment]
Fine by me.

**Comment by rhornung67** [comment]
The forked repo branch for this PR is pulled into the RAJA repo. The PR is here: https://github.com/LLNL/RAJA/pull/1177

**Pull request merged**